modules/spring-config/src/test/java/org/mule/runtime/config/internal/factories/FlowRefFactoryBeanTestCase.java

```diff
@@ -250,95 +272,182 @@ private Processor getFlowRefProcessor(FlowRefFactoryBean factoryBean) throws Exc
+    // MuleArtifactContext stubbing
+    DefaultListableBeanFactory beanFactory = new ObjectProviderAwareBeanFactory(null);
+    MuleArtifactContext muleArtifactContext = spy(createMuleArtifactContextStub(beanFactory));
+    // BeanFactory stubbing (subFlow and subFlow processor factories)
+    ComponentBuildingDefinition subFlowComponentBuildingDefinition = new CoreComponentBuildingDefinitionProvider()
+            .getComponentBuildingDefinitions()
+            .stream()
+            .filter(componentBuildingDefinition -> componentBuildingDefinition.getComponentIdentifier().getName().equals("sub-flow"))
+            .findFirst()
+            .get();
+    BeanDefinition subFlowProcessorBeanDefinition = genericBeanDefinition(Processor.class, () -> {
+      Processor subFlowProcessor = (Processor) mock(Object.class, INITIALIZABLE_MESSAGE_PROCESSOR);
+      when(subFlowProcessor.apply(any())).thenAnswer(successAnswer());
+      return subFlowProcessor;
+    }).getBeanDefinition();
+    BeanDefinition subFlowBeanDefinition = genericBeanDefinition(new ObjectFactoryClassRepository()
+            .getObjectFactoryClass(subFlowComponentBuildingDefinition, SubflowMessageProcessorChainFactoryBean.class, Object.class,
+                    () -> true, empty()))
+            .addPropertyValue("name", PARSED_DYNAMIC_REFERENCED_FLOW)
+            .addPropertyValue("messageProcessors", subFlowProcessorBeanDefinition)
+            .setScope(BeanDefinition.SCOPE_PROTOTYPE)
+            .getBeanDefinition();
+    beanFactory.registerBeanDefinition(PARSED_DYNAMIC_REFERENCED_FLOW, subFlowBeanDefinition);
+    //Additional flow and processing strategy (needed to generate a concurrent subflow instantiation)
+    Flow concurrentCallerFlow = mock(Flow.class, INITIALIZABLE_MESSAGE_PROCESSOR);
+    ProcessingStrategy concurrentCallerFlowProcessingStrategy = mock(ProcessingStrategy.class);
+    when(locator.find(Location.builder().globalName("concurrentFlow").build())).thenReturn(of(concurrentCallerFlow));
+    when(concurrentCallerFlow.getProcessingStrategy()).thenReturn(concurrentCallerFlowProcessingStrategy);
+    when(concurrentCallerFlowProcessingStrategy.onProcessor(any()))
+            .thenAnswer(invocationOnMock -> invocationOnMock.getArguments()[0]);
+    // Two flowRef dynamically pointing to the same subFlow
+    FlowRefFactoryBean flowRefFactoryBean = createFlowRefFactoryBean(DYNAMIC_REFERENCED_FLOW, "flow", muleArtifactContext);
+    FlowRefFactoryBean parallelFlowRefFactoryBean =
+            createFlowRefFactoryBean(DYNAMIC_REFERENCED_FLOW, "concurrentFlow", muleArtifactContext);
+    // Events are sent to both flowRefs in parallel in order to trigger a concurrent subflow instantiation
+    Thread flowEvents = new Thread(() -> {
+      sendEventsThroughFlowRef(flowRefFactoryBean);
+    }, "Flow Events");
+    Thread parallelFlowEvents = new Thread(() -> {
+      sendEventsThroughFlowRef(parallelFlowRefFactoryBean);
```

Comment:
can this be executed more times to force the manifestation of the race condition?

Reply:
Will try. With some modifications it might be possible to make it fail "almost all the executions".

Reply:
New version failed 1000 of 1000 runs without the fix